src/Tasks/Common/TaskToolBase.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    public abstract class ToolTaskBase : ToolTask
+    {
+        private Logger _logger;
+
+        internal ToolTaskBase(Logger logger = null)
+        {
+            _logger = logger;
+        }
+
+        internal new Logger Log
+        {
+            get
+            {
+                if (_logger == null)
+                {
+                    _logger = new LogAdapter(base.Log);
+                }
+
+                return _logger;
+            }
+        }
+
+        public override bool Execute()
```

Comment:
I don't think we should wholesale override ToolTask.Execute as base.Execute does a lot of things. Normally, you would override only things like ValidateParameters, GenerateFullPathToTool, GenerteCommandLineCommands and let Execute do its thing.  Crossgen is a little weird in that it takes two invocations ((1) to compile (2) to create profiling symbols), but it's hard for me to see what we're giving up by replacing Execute and calling ExecuteTool twice. I particularly don't like that we'd be baking this override into a base class for other tool invocations. I think we should probably split (1) and (2) into separate tasks.

Reply:
I need a custom implementation of Execute really, so I'll just remove the override from the base class (I agree with you... doesn't feel correct), and override it in my task.

Reply:
I'm still not sure about this. @rainersigwald do you think it's legitimate for a ToolTask to override Execute wholesale and use ExecuteTool multiple times? 

An alternative would be to have one task that produces the files to be crossgened, and chooses the crossgen, etc.  Then batch trough the list with a task that does the crossgening one execute at a time. And then do the profiling pdb generation in a separate batch.

Reply:
Why wouldn't it be legitimate for a ToolTask to override Execute and call ExecuteTool multiple times? The implementation of ToolTask.Execute seems to do a bunch of validations and command line building, which is what the crossgen task does here already, then calls ExecuteTool.

The implementation of ToolTask.ExecuteTool also seems capable of being invoked multiple times (https://github.com/Microsoft/msbuild/blob/master/src/Utilities/ToolTask.cs)

Reply:
It just doesn't feel like the way ToolTask was designed to be used.  Execute isn't particularly simple so it takes a lot of reading for me to see what we're giving up. For example, now I just saw that it handles the case where a tool has returned non-zero but didn't output anything that could be logged as an error and makes sure to log a generic error. We don't seem to have that handled.

Reply:
I can add a generic error. Based on my crossgen testing, that shouldn't really happen since crossgen has to emit some kind of failure message to stderr when it fails to compile an assembly

Reply:
We're missing other things from base Execute too:

Logging of command line (very useful when investigating issues from binlog).
Catching of failure to *start* the process.

It's tough to debate these 1 at a time. This feels like a case where we should follow the pattern and not have to worry about these details.

Reply:
Ping, @rainersigwald, I could use another opinion on this.

Reply:
If we must keep it this way, I think we can steal more code from base.Execute and use our own wrapper around ExecuteTool that mimics it more carefully.

Reply:
Thinking about this some more, I see other advantages to the split tasks approach I mentioned in https://github.com/nguerrera:

1. We can do incremental R2R easily and idiomatically using target batching and inputs and outputs
2. We can decouple the part that runs crossgen from the part that is tied to publish, allowing us to clean up dotnet store to share code with this
3. This decoupling will also let us use the product code to crossgen the product.

I can work on this as part of dotnet store cleanup that is currently on my plate and we can review separately. I appreciate that this review has gone on for a while and this is a significant refactoring.

Reply:
I appreciate @nguerrera's willingness to move this part of the code forward after this PR gets checked in. However, as a rule of thumb, we will take as long and as many iterations as it takes to get the code to a quality level that we feel comfortable with.
Experience has taught us that a short while after the code gets checked in, it is owned by us and I, in particular, don't have to incur too much debt right from out of the gates.
Just sharing our thought process and principles here.

Reply:
I agree that overriding `ToolTask.Execute` is surprising and likely to result in some unpleasant surprises. If it can be avoided, I think that's best--the "split into two tasks" approach sounds pretty good to me (but I don't have domain expertise in store/crossgen).

Reply:
I see, sounds reasonable. I'll look into refactoring this to not override ToolTask.Execute

Reply:
Refactored